File: src/main/java/com/team2502/ezauton/pathplanning/PathGenerator.java
package com.team2502.ezauton.pathplanning;

import com.team2502.ezauton.pathplanning.purepursuit.PPWaypoint;

import java.util.ArrayList;
import java.util.List;

public class PathGenerator
{

    private final PPWaypoint[] ppWaypoints;

    public PathGenerator(PPWaypoint... ppWaypoints)
    {
        this.ppWaypoints = ppWaypoints;
    }

    public Path generate()
    {
        List<PathSegment> pathSegments = new ArrayList<>();
        double distance = 0;
        for(int i = 0; i < ppWaypoints.length-1; i++)
        {
            PPWaypoint from = ppWaypoints[i];
            PPWaypoint to = ppWaypoints[i+1];
            double length = from.getLocation().dist(to.getLocation());

            //TODO: Update from RobotCode2018 style pathsegments
            PathSegment pathSegment = new PathSegment(from, to, i == 0, i == ppWaypoints.length - 2, distance, distance += length, length);
            pathSegments.add(pathSegment);
        }
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/pathplanning/PathGenerator.kt
package com.github.ezauton.core.pathplanning

import com.github.ezauton.conversion.Distance
import com.github.ezauton.conversion.LinearVelocity
import com.github.ezauton.conversion.Time
import com.github.ezauton.core.action.Speed
import com.github.ezauton.core.pathplanning.purepursuit.PPWaypoint
import java.io.Serializable
import java.util.*

class PathParams(val path: Path<Distance>, val speed: (Distance) -> LinearVelocity)

private typealias Segment = PathSegment<Distance>

/**
 * Generates a path 🗺 for Pure Pursuit given [PPWaypoint]
 */
class PathGenerator(private vararg val ppWaypoints: PPWaypoint) : Serializable {

  fun generate(dt: Time): PathParams {

    val interpolatorMap = HashMap<Double, SpeedInterpolator>()

    val segments = ArrayList<Segment>()
    for (i in 0 until ppWaypoints.size - 1) {
      val from = ppWaypoints[i]
      val to = ppWaypoints[i + 1]

      val line = LineSegment(from.location, to.location)
      segments.add(line)

      val length = line.length


      val beginningSpeed = if (i == 0 && from.speed.isZero) to.speed else from.speed
      val interpolator = SpeedInterpolator(length, beginningSpeed, to.speed, dt, from.acceleration, from.deceleration)

      interpolatorMap[length.value] = interpolator

    }

    val sortedMap = interpolatorMap.toSortedMap()

    val speed = TimeGetter(sortedMap)

    val path = Path(segments)

    return PathParams(path, speed)
  }
}

private class TimeGetter(val map: SortedMap<Double, SpeedInterpolator>) : Speed {

  private fun interpolator(distance: Distance): SpeedInterpolator {
    val tailMap = map.tailMap(distance.value)
    val key = if (tailMap.isEmpty()) map.lastKey() else tailMap.firstKey()
    return map[key]!!
  }

  override fun invoke(distance: Distance): LinearVelocity {
    val interp = interpolator(distance)
    return interp[distance]
  }

}
